### ISSUE TITLE
bpw calculation

### ISSUE THREAD

**Issue by Chainfire** [turboderp/exllamav2#37]
Is there an easy formula for calculating a target bpw to fill the VRAM but not OOM next to my desktop?

A 34b model at 4.0bpw is going to be ~17GB base, I guess. How much is the overhead to load? How much additional VRAM per 1k context? Scratch memory? Or is all that model-specific and/or is this just trial-and-error until satisfied?

Right now it seems like I have 2GB left on my 4090, though not sure if that would fill up if using 16k context.

**Comment by unphased** [comment]
yeah I wish there was a good reference somewhere to help us ballpark. I'm trying to do tests on a 12GB GPU and it is absolutely brutal, first I downloaded the 6.0, then 5.0, then 4.0, and then only the 3.0 quant of deepseek-coder-instruct-exl2 could even run without OOM.  I still think i'm doing something wrong though since the results are garbage and the amount of VRAM gobbled up looks like twice what it should be (i should be able to run something like a 3.0 quant of 13B?).

**Comment by turboderp** [comment]
deepseek-coder models all need the `-rs 4` argument to set the RoPE scale that the model expects. I'll add some code to read the relevant lines from the config automatically soon.

As for VRAM requirement, it's a very complicated to work out in advance. VRAM needs to hold model weights, except for the embedding layer, and then a number of temporary buffers to use during inference. The sizes of these buffers depend on the maximum number of tokens you'll want to be able process in a single batch. Then you need the K/V cache which scales with the hidden size of the model (reduced by GQA where applicable) times the context length and the depth. In addition, Torch and CUDA both add their own overhead, and then the question remains how much you need to reserve for anything else the GPU has to work on, like a desktop environment.

Overall it's just very difficult to give even a ballpark figure without measuring it for each version of every model. deepseek-coder-6.7B, for instance, uses 8 GB of VRAM for its full 16k context, whereas the 33B version needs only about 4.5 GB because it uses GQA. But I'm going to give it some thought and see if I can come up with some guidelines at least.

**Comment by unphased** [comment]
Thank you for that explanation which goes well over my head but certainly puts my concerns I mentioned to bed!! I did figure out the rope scaling value and got this particular model working on various systems in the two flavors, and it runs really fast... wow, that's a lot of memory consumed for context. 

Given how blazingly fast the 6.7B model runs for me, I wonder if the best way to use it from my 12GB machine is with a different runtime that might not consume 8GB just for its context. Or if there is a way to compromise on the context a bit at least to fit the 4.0 quant instead.

Btw the example chat terminal program is simply lovely.

**Comment by turboderp** [comment]
The size of the context is a feature of the model, not the runtime. For a model without GQA, the key/value store is always going to be the model's hidden_size \* num_layers \* 2 \* sizeof(datatype) \* max_seq_len. In this case that's 4096 \* 32 \* 2 \* 2 \* 16384 =  8 * 1024^3 bytes.

There's not a lot you can really do about that. You could limit the context size if you don't need all 16k tokens, or you could use the 8-bit cache mode which would cut the cache size in half. But you also lose some precision that way and possibly more than you'd gain by going from 3 to 4 bit weights.

There are other models you could try, though. Mistral-7B and anything derived from it could be worth considering, since aside from just being a very good base model, Mistral uses GQA with 8 K/V heads vs 32 Q heads, which reduces the cache size to a quarter. There's [MistralLite](https://huggingface.co/amazon/MistralLite) for instance which should natively support 16k context and would only need 2 GB of VRAM for it.

**Comment by unphased** [comment]
Thank you for the suggestion. I can indeed run MistralLite 8bpw on this machine (just barely). All these 7b models execute extremely quickly on the 3080Ti! I noticed when running deepseek coder 33b 4.0 on a 3090 that ram usage sawtooths from 21 or so to nearly 24 (99% mem util) but doesn't OOM. Should I assume this is a sort of periodic "garbage collection" and that if I had less free vram headroom I could merely expect that gc to run more frequently?

For clarification, if I truncate the context to half in that situation, to gain back 4GB (wow thats a lot), would this also degrade precision or is that aspect only in reference to "8-bit cache mode" for context?

Thank you so much for your assistance. I hope to contribute to your project in some way soon.

**Comment by unphased** [comment]
Ah ok i think it works, context size is controllable with just the length parameter. i tried a smaller one and vram usage dropped enormously on this non GQA deepseek 6.7b.
